File: backend/app/api/v1/artifact_search.py
"""Artifact Search API - global search across all artifacts."""

from datetime import datetime
from fastapi import APIRouter, Query

from app.runtime.manager import get_runtime
from app.utils.logger import get_logger

logger = get_logger(__name__)

router = APIRouter(tags=["artifact-search"])


@router.get("/artifacts/search")
async def search_artifacts(
    q: str | None = Query(None, description="Search query (name/content)"),
    type: str | None = Query(None, description="Filter by artifact type"),
    agent_id: str | None = Query(None, description="Filter by agent ID"),
    task_id: str | None = Query(None, description="Filter by task ID"),
    workspace_id: str | None = Query(None, description="Filter by workspace ID"),
    step_id: str | None = Query(None, description="Filter by step ID"),
    created_after: datetime | None = Query(None, description="Filter by created_after"),
    created_before: datetime | None = Query(None, description="Filter by created_before"),
    sort_by: str = Query("created_at", description="Sort field: created_at|name|type"),
    order: str = Query("desc", description="Sort order: asc|desc"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
):
    """Search artifacts with filters, sorting, and pagination."""
    runtime = get_runtime()

    # Validate sort_by
    if sort_by not in ("created_at", "name", "type"):
        sort_by = "created_at"
    if order not in ("asc", "desc"):
        order = "desc"

    items, total = runtime.workspace_manager.search_items(
        query=q,
        artifact_type=type,
        agent_id=agent_id,
        task_id=task_id,
        workspace_id=workspace_id,
        step_id=step_id,
        created_after=created_after,
        created_before=created_before,
        sort_by=sort_by,
        order=order,
        limit=limit,
        offset=offset,
    )

    return {
        "items": items,
        "total": total,
        "limit": limit,
        "offset": offset,
    }

File: frontend/src/api/artifactSearch.ts
import apiClient from "./client"
import type { ArtifactItem } from "./artifacts"

export interface ArtifactSearchParams {
  q?: string
  type?: string
  agent_id?: string
  task_id?: string
  workspace_id?: string
  step_id?: string
  created_after?: string
  created_before?: string
  sort_by?: "created_at" | "name" | "type"
  order?: "asc" | "desc"
  limit?: number
  offset?: number
}

export interface ArtifactSearchResponse {
  items: ArtifactItem[]
  total: number
  limit: number
  offset: number
}

export async function searchArtifacts(params: ArtifactSearchParams): Promise<ArtifactSearchResponse> {
  const clean: Record<string, string | number> = {}
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null && v !== "") {
      clean[k] = v
    }
  }
  const { data } = await apiClient.get("/artifacts/search", { params: clean })
  return data
}